### ISSUE TITLE
Why NEGOTIATE_ALLREDUCE is much longer in TensorFlow comparing to PyTorch?

### ISSUE THREAD

**Issue by zarzen** [horovod/horovod#1454]
**Environment:**
1. Framework: TensorFlow, Keras, PyTorch
2. Framework version: tensorflow -1.14.0, torch-1.3.0
3. Horovod version: 0.18.1
4. MPI version: openmpi-4.0.2
5. CUDA version: None
6. NCCL version: None
7. Python version: 3.7.0
8. OS and version: Ubuntu 16.04
9. GCC version: 5.4.0

**Your question:**
I am running the experiments on two machines no GPU. The experiment scripts are from [horovod/examples](https://github.com/horovod/horovod/tree/master/examples), `pytorch_mnist.py` and `tensorflow_mnist.py` respectively. 
I recorded the timeline-file, based on the trace, I found it takes 124ms for the TensorFlow experiment to complete "NEGOTIATE_ALLREDUCE". While it only takes 24ms for PyTorch. Any reason for this phenomenon? Detailed timelines are listed, timeline files are at [tf-timeline](https://drive.google.com/file/d/1A2OZl72Ris36Gxk017nZRKmBD38jN6Tv/view?usp=sharing) and [torch-timeline](https://drive.google.com/file/d/1TsTbcRt1mYrf31ORN7X8ph0xwwbxDngo/view?usp=sharing)

## TensorFlow records
Name | Wall Duration | Self time| Average Wall Duration| Occurences
-- | -- | -- | -- | --
NEGOTIATE_ALLREDUCE 🔍 | 156,489.507 ms | 156,489.507 ms | 125.192 ms | 1250
ALLREDUCE 🔍 | 229,407.956 ms | 869.626 ms | 183.526 ms | 1250
INIT_FUSION_BUFFER 🔍 | 0.041 ms | 0.041 ms | 0.041 ms | 1
MEMCPY_IN_FUSION_BUFFER 🔍 | 5,757.987 ms | 5,757.987 ms | 4.606 ms | 1250
MPI_ALLREDUCE 🔍 | 218,325.188 ms | 218,325.188 ms | 174.660 ms | 1250
MEMCPY_OUT_FUSION_BUFFER 🔍 | 4,455.114 ms | 4,455.114 ms | 3.564 ms | 1250

## PyTorch

Name | Wall Duration | Self time| Average Wall Duration| Occurences
-- | -- | -- | -- | --
NEGOTIATE_ALLREDUCE 🔍 | 13,088.435 ms | 13,088.435 ms | 22.184 ms | 590
ALLREDUCE 🔍 | 528.083 ms | 34.663 ms | 0.895 ms | 590
MEMCPY_IN_FUSION_BUFFER 🔍 | 4.171 ms | 4.171 ms | 0.007 ms | 585
MPI_ALLREDUCE 🔍 | 487.103 ms | 487.103 ms | 0.826 ms | 590
MEMCPY_OUT_FUSION_BUFFER 🔍 | 2.146 ms | 2.146 ms | 0.004 ms | 585

## MXNet 
Name | Wall Duration | Self time| Average Wall Duration| Occurences
-- | -- | -- | -- | --
NEGOTIATE_ALLREDUCE 🔍 | 39,862.220 ms | 39,862.220 ms | 137.456 ms | 290
ALLREDUCE 🔍 | 50,895.010 ms | 204.333 ms | 175.500 ms | 290
INIT_FUSION_BUFFER 🔍 | 0.018 ms | 0.018 ms | 0.018 ms | 1
MEMCPY_IN_FUSION_BUFFER 🔍 | 196.994 ms | 196.994 ms | 0.679 ms | 290
MPI_ALLREDUCE 🔍 | 50,189.702 ms | 50,189.702 ms | 173.068 ms | 290
MEMCPY_OUT_FUSION_BUFFER 🔍 | 303.963 ms | 303.963 ms | 1.048 ms | 290

Trace file: 
[mxnet-timelines](https://drive.google.com/file/d/1aBXJl75tQ6T54InnIAFrvY2ZYdnBkC5h/view?usp=sharing)

## Question
not only `NEGOTIATE_ALLREDUCE` times are significantly different, but also `ALLREDUCE`. I have checked two models are very similar. Even though in `tensorflow_mnist.py` optimizer is `Adam`, but I think processes only communicate with gradients? then their sizes should be similar to each other.

**Comment by tgaddair** [comment]
Thanks for the detailed investigation @zarzen!  @abditag2 is going to look into this on our side and see if we can reproduce / identify the source of the discrepancy.

**Comment by karan6181** [comment]
@zarzen Could you please also try and post your result with MXNet-1.5.1 or MXNet-1.5.0. Since you have the similar environment, it would be easy to compare across different frameworks.

MXNet Installation: https://mxnet.apache.org/get_started?version=v1.5.1&platform=linux&language=python&environ=pip&processor=cpu

Thanks!

**Comment by zarzen** [comment]
@karan6181 Hi, I just uploaded the tracing for mxnet, hope that helps.

**Comment by lxl910915** [comment]
Same issue. Allreducing a tensor with shape [1024, 1024], HorovodAllreduce takes 9ms, but pytorch 0.02ms.

**Comment by lxl910915** [comment]
We open HOROVOD_LOG_LEVEL=trace, and get that the process from Allreduce enqueued to response takes about 4ms. Three examples:
```

[2019-10-30 18:33:07.224617: T horovod/common/operations.cc:766] [1]: Enqueued bert/encoder/layer_22/output/HorovodAllreduce_bert_encoder_layer_22_output_mul_3_0
[2019-10-30 18:33:07.228613: D horovod/common/controller.cc:669] Created response of size 4194304
[2019-10-30 18:33:07.228624: T horovod/common/controller.cc:278] Sending ready responses as bert/encoder/layer_22/output/HorovodAllreduce_bert_encoder_layer_22_output_mul_3_0;
[2019-10-30 18:33:07.228661: T horovod/common/operations.cc:532] [1]: Performing bert/encoder/layer_22/output/HorovodAllreduce_bert_encoder_layer_22_output_mul_3_0
[2019-10-30 18:33:07.228670: D horovod/common/operations.cc:533] [1]: Processing 1 tensors
[2019-10-30 18:33:07.228795: T horovod/common/operations.cc:536] [1]: Finished performing bert/encoder/layer_22/output/HorovodAllreduce_bert_encoder_layer_22_output_mul_3_0




[2019-10-30 18:33:07.22307: T horovod/common/operations.cc:766] [0]: Enqueued bert/encoder/layer_2/output/HorovodAllreduce_bert_encoder_layer_2_output_mul_3_0
[2019-10-30 18:33:07.26350: D horovod/common/controller.cc:669] Created response of size 4194304
[2019-10-30 18:33:07.26360: T horovod/common/controller.cc:278] Sending ready responses as bert/encoder/layer_2/output/HorovodAllreduce_bert_encoder_layer_2_output_mul_3_0;
[2019-10-30 18:33:07.26396: T horovod/common/operations.cc:532] [0]: Performing bert/encoder/layer_2/output/HorovodAllreduce_bert_encoder_layer_2_output_mul_3_0
[2019-10-30 18:33:07.26406: D horovod/common/operations.cc:533] [0]: Processing 1 tensors
[2019-10-30 18:33:07.26519: T horovod/common/operations.cc:536] [0]: Finished performing bert/encoder/layer_2/output/HorovodAllreduce_bert_encoder_layer_2_output_mul_3_0



[2019-10-30 18:33:06.355048: T horovod/common/operations.cc:766] [1]: Enqueued bert/encoder/layer_7/attention/self/output/HorovodAllreduce_bert_encoder_layer_7_attention_self_output_dense_BiasAdd_0
[2019-10-30 18:33:06.358979: D horovod/common/controller.cc:669] Created response of size 4194304
[2019-10-30 18:33:06.358989: T horovod/common/controller.cc:278] Sending ready responses as bert/encoder/layer_7/attention/self/output/HorovodAllreduce_bert_encoder_layer_7_attention_self_output_dense_BiasAdd_0;
[2019-10-30 18:33:06.359026: T horovod/common/operations.cc:532] [1]: Performing bert/encoder/layer_7/attention/self/output/HorovodAllreduce_bert_encoder_layer_7_attention_self_output_dense_BiasAdd_0
[2019-10-30 18:33:06.359036: D horovod/common/operations.cc:533] [1]: Processing 1 tensors
[2019-10-30 18:33:06.359151: T horovod/common/operations.cc:536] [1]: Finished performing bert/encoder/layer_7/attention/self/output/HorovodAllreduce_bert_encoder_layer_7_attention_self_output_dense_BiasAdd_0   
```

**Comment by lxl910915** [comment]
In RunLoopOnce() method, there is a sleep() method. but the sleep time is too long, it is 5ms - A. And A is one loop time. In our case, one loop time is about 1ms, so it will sleep 4ms. We change the sleep time to 10 nanoseconds, the performance increase about 80%.
```
bool RunLoopOnce(HorovodGlobalState& state) {
  // This delay determines thread frequency and communication message latency
  auto start_time = std::chrono::steady_clock::now();
  auto sleep_duration = state.last_cycle_start +
                        std::chrono::microseconds(long(
                            state.parameter_manager.CycleTimeMs() * 1000.)) -
                        start_time;
  LOG(TRACE) << "sleep_duration " << sleep_duration.count();
  if (sleep_duration > std::chrono::steady_clock::duration::zero()) {
    std::this_thread::sleep_for(sleep_duration); // std::chrono::nanoseconds(10)
  }
  LOG(TRACE) << "after sleep_duration ";
  ...
```

If you do not want to build horovod, there is a env HOROVOD_CYCLE_TIME to change the sleep time. HOROVOD_CYCLE_TIME's default value is 5.

```
   // Override the cycle time.
  state.parameter_manager.SetCycleTimeMs(5);
  auto horovod_cycle_time = std::getenv(HOROVOD_CYCLE_TIME);
  if (horovod_cycle_time != nullptr) {
    state.parameter_manager.SetCycleTimeMs(
        std::strtof(horovod_cycle_time, nullptr), true);
  }
```

**Comment by tgaddair** [comment]
Good find @lxl910915!  I would also suggest giving our [autotuner](https://github.com/horovod/horovod/blob/master/docs/autotune.rst) a try.  It will experiment with different values of the cycle time and fusion buffer to find the optimal values in your environment.  You can run it with `horovodrun --autotune` or `mpirun -x HOROVOD_AUTOTUNE=1`.

FYI, the cycle time can also be set in horovodrun with `horovodrun --cycle-time-ms`.

**Comment by tgaddair** [comment]
@zarzen, I would suggest giving the autotuner a try and seeing if that improves performance for TensorFlow.  Also, @romerojosh had mentioned that a possible cause for the different in `NEGOTIATE_ALLREDUCE` times might be due to the difference in when TensorFlow vs PyTorch submit tensors for allreduce, where one may be waiting for tensors to be ready while the other does not.

**Comment by lxl910915** [comment]
> Good find @lxl910915! I would also suggest giving our [autotuner](https://github.com/horovod/horovod/blob/master/docs/autotune.rst) a try. It will experiment with different values of the cycle time and fusion buffer to find the optimal values in your environment. You can run it with `horovodrun --autotune` or `mpirun -x HOROVOD_AUTOTUNE=1`.
> 
> FYI, the cycle time can also be set in horovodrun with `horovodrun --cycle-time-ms`.

We have tried `-x HOROVOD_AUTOTUNE=1 -x HOROVOD_AUTOTUNE_LOG=horovod_tune.log`. It will takes a long time to converge.

**Comment by tgaddair** [comment]
Hey @lxl910915, is that 4 hours to finish autotuning, or 4 hours to train the model?  How long does it take to converge without autotuning?

**Comment by lxl910915** [comment]
> Hey @lxl910915, is that 4 hours to finish autotuning, or 4 hours to train the model? How long does it take to converge without autotuning?

4 hours to finish autotuning.  @tgaddair

**Comment by whatdhack** [comment]
@zarzen  @tgaddair, were all the numbers taken with tf-timeline enabled ?  Tf-timelibe  slows down processing  by 10x.   Was the reason for  large  numbers for  Tensorflow ?
